README.md
[![sselib.js](http://dl.dropboxusercontent.com/u/15640279/massforstroelse-site/sse-lib.png)](https://npmjs.org/package/sselib)

_SSE (Server Sent Events) library for node.js._

`sselib` is a implementation of the server-side part of the [SSE] [1] protocol written in `Coffee Script` for the `node.js` platform.

[![Build Status](https://travis-ci.org/massforstroelse/sselib.js.png)](https://travis-ci.org/massforstroelse/sselib.js)

  [1]: http://dev.w3.org/html5/eventsource/

## Installation ##

### Install with npm: ###

    $ npm install sselib

### Requirements ###

None, tested on node.js 0.6 >

## Connect and Express Middleware ##

`sselib` can be used as a middleware for applications following the Connect convention.

### Example ###

#### Javascript ####

```javascript
    var sselib = require('sselib'),
    express = require('express');

    var app = express();
    
    app.use(sselib.middleware());
    
    app.get('/events', function(req, res) {
      res.sse({
        event: 'update',
        data: 'I am a stray cat.'
      });
        
    });

    app.listen(3000);
```

#### Coffeescript ####

```coffeescript
    sselib = require 'sselib'
    express = require 'express'

    app = express()

    app.use sselib.middleware()
    
    app.get '/events', (req, res) ->
      res.sse
        event: 'update'
        data: 'I am a stray cat.'
```

### Options ###

You can pass options when initializing the middleware.

```javascript
    app.use(sselib.middleware({
      retry: 5*1000,
      keepAlive: 15*1000,
      compatibility: true
    });
```

A already initialized connection can use `res.sse.set` in order to set a new value for the option. Setting longer retry values and closing connections can be a effective strategy when servers are under high load.

```javascript
    app.get('/events', function(req, res) {
      res.sse.set('retry', 30*1000);
    });
```

`res.sse.get(option)` respectively gets the current configuration.

#### retry ####

The time in milliseconds for client reconnects. Default is 5 seconds.  
Set to `false` in order to disable.

#### keepAlive ####

Sends pseudo keep alive heartbeats in order to keep the connection open. The value is the amount of milliseconds between each keepAlive heartbeat. Default is 15 seconds.  
Set to `false` in order to disable.

#### compatibility ####

"Quirks mode". Adds support for some polyfills and the way MSIE handles XDomainRequest. Default is `true`  
Set to `false` in order to disable.

## Use as a library to serialize data for your own transmission ##

### Example ###

```javascript
    var sselib = require('sselib');

    console.log(sselib.event("notice")); // "event: notice\n"
    console.log(sselib.data("Hello there!")); // "data: Hello there!\n\n"

    // or:
    
    sselib.data("Hello there!", function(err, result) {
      if (err) {
        // print the error to console
        console.log(err);
        return;
      }
      console.log(result) // "data: Hello there!\n\n"
    });
```

### Serializers ###

#### sselib.comment(comment [, callback]) ####

Returns a SSE-serialized comment `string` representing a comment (please note that comments are invisible to browser clients).

#### sselib.event(event [, callback]) ####

Returns a SSE-serialized event `string` that can be used with a following `data` type to trigger a event in the browser.

#### sselib.id(id [, callback]) ####

Returns a SSE-serialized id `string`. If called without `id` it will use a `UNIX timestamp` as the `id`.

#### sselib.data(data [, callback]) ####

Returns a SSE-serialized data `string`.

#### sselib.message(obj [, callback]) ####

`message` is provided as a meta-serializer. It will return a SSE-serialized string from a message object you pass in.

#### sselib.headers([callback]) ####

Returns a `Object` containing valid HTTP-headers suitable for a `http.ServerResponse`.

## Concepts ##

### Message Object ###

[![sselib.message() graph](http://dl.dropboxusercontent.com/u/15640279/massforstroelse-site/sselib-serialization-graph.png)](https://npmjs.org/package/sselib)

A `message object` is simply a javascript object containing the `data` and `event` keys, it can also optionally be given a `id` key.

#### Example ####

    {event: 'update', data: 'I am a stray cat.', id: 789}

## License ##

BSD (see [LICENSE.md](https://github.com/massforstroelse/sselib.js/blob/master/LICENSE.md "LICENSE.md"))  

Made by [massförströelse](http://massforstroel.se/ "massförströel.se")  



changelog.md
0.0.95
======

  * Retry is sent after compat-actions in order to not send headers prematurely.


0.0.94
======

  * Transfer-Encoding is set to chunked in compatibility mode.

0.0.93
======

  * Double linefeed sent after processed message.

0.0.92
======

  * Normal message dispatching now writes an extra linefeed after data to trigger the event.

0.0.91
======

  * Hot fix for headers not being sent when using yaffles polyfill.

0.0.9
=====

  * Added support for lastEventID as querystring.

0.0.8
=====

  * Added _socket to middleware.

0.0.7
=====

  * Using response.setHeader instead of writing them out.

0.0.6
=====

  * Correctified the repo url.
  * Changed name of changelog.

0.0.5
=====

  * Refactoring middleware to only expose middleware specific properties.
  * Added toString serialization.
  * Making keepalive timer private.
  * Adding module.exports.utils ns.
  * Middleware only calls next() if available.

0.0.4
=====

  * Added express as a dev dep.
  * Added tests for middleware instance.
  * Placing instance vars directly under res.sse for middleware.
  * Implemented get() and set().

0.0.3
=====

  * Added options.compatibility, this closes [#9](https://github.com/massforstroelse/sselib.js/issues/9)
  * Retry can be set to false, then no initial retry message will be sent to the client. This closes [#5](https://github.com/massforstroelse/sselib.js/issues/5).
  * _ prepending internal utils.
  * Changed internal interface for middleware.
  * Added graph to readme.
  * Data now splits to multiple data-lines on linefeed.

0.0.2
=====

  * Added logo to docs, this closes [#8](https://github.com/massforstroelse/sselib.js/issues/8).
  * Changed extend implemention in order to target node =< .6

0.0.1
=====

  * Implementing error as first arg in callbacks.
  * Added callbacks to message types.
  * Using versions instead of latest in npm deps.
  * Added req, res and options to testcase.
  * Making better use of header setting, this closes [#2](https://github.com/massforstroelse/sselib.js/issues/2).

0.0.1-beta
==========

  * First release.


test/test.js
// Generated by CoffeeScript 1.6.2
var MIDDLEWARE_INSTANCE_PROPERTIES, Mock, SOCKET_INSTANCE_ALIASES, SOCKET_INSTANCE_OPTIONS_KEYS, SOCKET_INSTANCE_PROPERTIES_PRIVATE, SOCKET_INSTANCE_PROPERTIES_PUBLIC, connect, connectApp, express, expressApp, request, sendTestMessage, sselib, testMessage, testMiddleware;

request = require('supertest');

connect = require('connect');

express = require('express');

sselib = require('../sselib');

Mock = function() {
  var _this = this;

  this.req = {};
  this.req.url = "http://example.com/";
  this.req.headers = {};
  this.req.headers['last-event-id'] = "keyboard-cat";
  this.req.headers.accept = 'text/event-stream';
  this.req.socket = {};
  this.req.socket.address = function() {
    return {
      port: 12346,
      family: 'IPv4',
      address: '127.0.0.1'
    };
  };
  this.res = {};
  this.res.headers = {};
  this.res.setHeader = function(k, v) {
    return _this.res.headers[k] = v;
  };
  this.res.writeHead = function(code, reason, headers) {};
  this.res.once = function() {};
  this.res.end = function() {};
  this.res.write = function(chunk, encoding) {};
  return this;
};

testMessage = {
  id: 42,
  event: "hi",
  data: "yo"
};

SOCKET_INSTANCE_PROPERTIES_PUBLIC = ['sendComment', 'sendRetry', 'sendEvent', 'sendId', 'sendData', 'sendRaw', 'req', 'res', 'options', 'set', 'get', 'toString'];

SOCKET_INSTANCE_PROPERTIES_PRIVATE = ['_processAndSendMessage', '_dispatchMessage', '_writeHeaders', '_keepAlive', '_compatibility'];

SOCKET_INSTANCE_ALIASES = ['pub', 'publish', 'send'];

SOCKET_INSTANCE_OPTIONS_KEYS = ['keepAlive', 'retry', 'compatibility'];

MIDDLEWARE_INSTANCE_PROPERTIES = ['sendComment', 'sendRetry', 'sendEvent', 'sendId', 'sendData', 'sendRaw', 'set', 'get', 'toString', 'send'];

describe('SSE', function() {
  describe('comment()', function() {
    it('should return a valid comment', function(done) {
      sselib.comment('cat').should.equal(': cat\n\n');
      return done();
    });
    return it('should return a valid comment to a optional callback', function(done) {
      return sselib.comment('cat', function(err, result) {
        result.should.equal(': cat\n\n');
        return done();
      });
    });
  });
  describe('retry()', function() {
    it('should return a valid retry instruction', function(done) {
      sselib.retry(3000).should.equal('retry: 3000\n');
      return done();
    });
    return it('should return a valid retry instruction to a optional callback', function(done) {
      sselib.retry(3000, function(err, result) {
        return result.should.equal('retry: 3000\n');
      });
      return done();
    });
  });
  describe('event()', function() {
    it('should return a valid event record', function(done) {
      sselib.event('cat').should.equal('event: cat\n');
      return done();
    });
    return it('should return a valid event record to a optional callback', function(done) {
      sselib.event('cat', function(err, result) {
        return result.should.equal('event: cat\n');
      });
      return done();
    });
  });
  describe('id()', function() {
    it('should return a valid id record', function(done) {
      sselib.id('cat').should.equal('id: cat\n');
      return done();
    });
    return it('should return a valid id record to a optional callback', function(done) {
      sselib.id('cat', function(err, result) {
        return result.should.equal('id: cat\n');
      });
      return done();
    });
  });
  describe('data()', function() {
    it('should return a valid data record', function(done) {
      sselib.data('cat').should.equal('data: cat\n\n');
      return done();
    });
    it('should return a valid data record to a optional callback', function(done) {
      sselib.data('cat', function(err, result) {
        return result.should.equal('data: cat\n\n');
      });
      return done();
    });
    it('should return a valid data record', function(done) {
      sselib.data('cat\nmouse').should.equal('data: cat\ndata: mouse\n\n');
      return done();
    });
    return it('should return a valid data record to a optional callback', function(done) {
      sselib.data('cat\nmouse', function(err, result) {
        return result.should.equal('data: cat\ndata: mouse\n\n');
      });
      return done();
    });
  });
  describe('message()', function() {
    it('should serialize a JSON to valid SSE', function(done) {
      sselib.message(testMessage).should.equal('id: 42\nevent: hi\ndata: yo\n\n');
      return done();
    });
    return it('should serialize a JSON to valid SSE to a optional callback', function(done) {
      sselib.message(testMessage, function(err, result) {
        return result.should.equal('id: 42\nevent: hi\ndata: yo\n\n');
      });
      return done();
    });
  });
  return describe('headers()', function() {
    it('should return a object containing valid SSE Headers', function(done) {
      sselib.headers().should.eql({
        'Content-Type': 'text/event-stream; charset=utf-8',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive',
        'Transfer-Encoding': 'identity'
      });
      return done();
    });
    return it('should return a object containing valid SSE Headers to a optional callback', function(done) {
      sselib.headers(function(err, result) {
        return result.should.eql({
          'Content-Type': 'text/event-stream; charset=utf-8',
          'Cache-Control': 'no-cache',
          'Connection': 'keep-alive',
          'Transfer-Encoding': 'identity'
        });
      });
      return done();
    });
  });
});

describe('Initialized SSE', function() {
  describe('The toString serialization should work', function() {
    return it('should serialize properly', function(done) {
      var instance, mock;

      mock = new Mock();
      instance = new sselib(mock.req, mock.res);
      instance.toString().should.equal("<SSE 127.0.0.1:12346 (IPv4)>");
      return done();
    });
  });
  describe('The @options should be populated with default values', function() {
    var instance, mock;

    mock = new Mock();
    instance = new sselib(mock.req, mock.res);
    it('should have all the keys', function(done) {
      instance.options.should.have.keys(SOCKET_INSTANCE_OPTIONS_KEYS);
      return done();
    });
    it('should have the default value for keepAlive', function(done) {
      instance.options.should.have.property('keepAlive', 15 * 1000);
      return done();
    });
    it('should have the default value for retry', function(done) {
      instance.options.should.have.property('retry', 5 * 1000);
      return done();
    });
    return it('should have the default value for compatibility', function(done) {
      instance.options.should.have.property('compatibility', true);
      return done();
    });
  });
  describe('The options passed should take effect on the instance', function() {
    var instance, mock, options;

    options = {
      keepAlive: false,
      retry: 10 * 1000,
      compatibility: false
    };
    mock = new Mock();
    instance = new sselib(mock.req, mock.res, options);
    it('should have all the keys', function(done) {
      instance.options.should.have.keys(SOCKET_INSTANCE_OPTIONS_KEYS);
      return done();
    });
    it('should have the passed value for keepAlive', function(done) {
      instance.options.should.have.property('keepAlive', false);
      return done();
    });
    it('should have the passed value for retry', function(done) {
      instance.options.should.have.property('retry', 10 * 1000);
      return done();
    });
    return it('should have the passed value for compatibility', function(done) {
      instance.options.should.have.property('compatibility', false);
      return done();
    });
  });
  describe('The getter method should work', function() {
    var instance, mock, options;

    options = {
      keepAlive: false,
      retry: 10 * 1000,
      compatibility: false
    };
    mock = new Mock();
    instance = new sselib(mock.req, mock.res, options);
    it('should be able to get keepAlive', function(done) {
      instance.get('keepAlive').should.eql(false);
      return done();
    });
    it('should be able to get retry', function(done) {
      instance.get('retry').should.eql(10 * 1000);
      return done();
    });
    return it('should be able to get compatibility', function(done) {
      instance.get('compatibility').should.eql(false);
      return done();
    });
  });
  describe('The setter method should work', function() {
    var instance, mock;

    mock = new Mock();
    instance = new sselib(mock.req, mock.res);
    it('should be able to set keepAlive', function(done) {
      instance.set('keepAlive', 5000);
      instance.get('keepAlive').should.eql(5000);
      return done();
    });
    it('should be able to set retry', function(done) {
      instance.set('retry', 20000);
      instance.get('retry').should.eql(20 * 1000);
      return done();
    });
    return it('should be able to set compatibility', function(done) {
      instance.set('compatibility', true);
      instance.get('compatibility').should.eql(true);
      return done();
    });
  });
  describe('The socket object should have all the public properties', function() {
    var instance, mock;

    mock = new Mock();
    instance = new sselib(mock.req, mock.res);
    return SOCKET_INSTANCE_PROPERTIES_PUBLIC.forEach(function(property) {
      return it("should have " + property, function(done) {
        instance.should.have.property(property);
        return done();
      });
    });
  });
  describe('The socket object should have all the private properties', function() {
    var instance, mock;

    mock = new Mock();
    instance = new sselib(mock.req, mock.res);
    return SOCKET_INSTANCE_PROPERTIES_PRIVATE.forEach(function(property) {
      return it("should have " + property, function(done) {
        instance.should.have.property(property);
        return done();
      });
    });
  });
  return describe('The socket object should have all the aliases', function() {
    var instance, mock;

    mock = new Mock();
    instance = new sselib(mock.req, mock.res);
    return SOCKET_INSTANCE_ALIASES.forEach(function(property) {
      return it("should have " + property, function(done) {
        instance.should.have.property(property);
        return done();
      });
    });
  });
});

testMiddleware = function(connectApp, expressApp) {
  return describe("As middleware", function() {
    describe('As connect middleware', function() {
      return it('should respond and attach itself whenever seeing event-stream accept headers', function(done) {
        return request(connectApp).get('/').set('Accept', 'text/event-stream').expect(200).expect('Content-Type', /text\/event-stream/).end(function(err, res) {
          if (err) {
            return done(err);
          }
          return done();
        });
      });
    });
    describe('As express middleware', function() {
      return it('should respond and attach itself whenever seeing event-stream accept headers', function(done) {
        return request(expressApp).get('/').set('Accept', 'text/event-stream').expect(200).expect('Content-Type', /text\/event-stream/).end(function(err, res) {
          if (err) {
            return done(err);
          }
          return done();
        });
      });
    });
    return describe('middleware API', function() {
      var app, mock;

      app = sselib.middleware();
      mock = new Mock();
      it('should provide the sse property to res', function(done) {
        return app(mock.req, mock.res, function() {
          mock.res.should.have.property('sse');
          return done();
        });
      });
      return MIDDLEWARE_INSTANCE_PROPERTIES.forEach(function(property) {
        return it("should provide the public middleware property " + property + " under res.sse", function(done) {
          return app(mock.req, mock.res, function() {
            mock.res.sse.should.have.property(property);
            return done();
          });
        });
      });
    });
  });
};

sendTestMessage = function(req, res, next) {
  var _this = this;

  res.sse(testMessage);
  setTimeout((function() {
    return res.end();
  }), 2 * 1000);
  return next();
};

connectApp = connect();

connectApp.use(sselib.middleware());

expressApp = express();

expressApp.use(sselib.middleware());

testMiddleware(connectApp, expressApp);


sselib.js
// Generated by CoffeeScript 1.6.2
var EventEmitter, MIDDLEWARE_INSTANCE_PROPERTIES, SSE, error, middleware, url, _utils,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

EventEmitter = require('events').EventEmitter;

url = require('url');

error = null;

_utils = {};

_utils.typeCheck = function(type, obj) {
  var cls;

  cls = Object.prototype.toString.call(obj).slice(8, -1);
  return obj !== void 0 && obj !== null && cls === type;
};

_utils.extend = function(origin, extension) {
  var key, value;

  if (!extension || !_utils.typeCheck('Object', extension)) {
    return origin;
  }
  for (key in extension) {
    value = extension[key];
    if (origin[key] == null) {
      origin[key] = value;
    }
  }
  return origin;
};

module.exports.utils = _utils;

SSE = (function(_super) {
  __extends(SSE, _super);

  SSE.defaultOptions = {
    retry: 5 * 1000,
    keepAlive: 15 * 1000,
    compatibility: true
  };

  SSE.comment = function(comment, callback) {
    var serialized;

    serialized = ": " + comment + "\n\n";
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.retry = function(time, callback) {
    var serialized;

    serialized = "retry: " + time + "\n";
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.event = function(event, callback) {
    var serialized;

    serialized = event ? "event: " + event + "\n" : '';
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.id = function(id, callback) {
    var serialized;

    if (_utils.typeCheck('Function', id)) {
      callback = id;
      id = null;
    }
    serialized = "id: " + (id ? id : (new Date()).getTime()) + "\n";
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.data = function(data, callback) {
    var piece, serialized, _i, _len;

    serialized = '';
    if (!(_utils.typeCheck('String', data) && (data != null))) {
      data = JSON.stringify(data);
      serialized = data ? "data: " + data + "\n" : '';
    } else {
      data = data.split('\n');
      for (_i = 0, _len = data.length; _i < _len; _i++) {
        piece = data[_i];
        serialized += "data: " + piece + "\n";
      }
    }
    if (serialized) {
      serialized += '\n';
    }
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.message = function(obj, callback) {
    var serialized;

    serialized = [this.id(obj.id), this.event(obj.event), this.data(obj.data)].join('');
    if (!callback) {
      return serialized;
    } else {
      return callback(error, serialized);
    }
  };

  SSE.headers = function(callback) {
    var headerDict;

    headerDict = {
      'Content-Type': 'text/event-stream; charset=utf-8',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
      'Transfer-Encoding': 'identity'
    };
    if (!callback) {
      return headerDict;
    } else {
      return callback(error, headerDict);
    }
  };

  function SSE(req, res, options) {
    var _this = this;

    this.req = req;
    this.res = res;
    this.options = options != null ? options : {};
    this._dispatchMessage = __bind(this._dispatchMessage, this);
    this.sendRaw = __bind(this.sendRaw, this);
    this.sendData = __bind(this.sendData, this);
    this.sendId = __bind(this.sendId, this);
    this.sendEvent = __bind(this.sendEvent, this);
    this.sendRetry = __bind(this.sendRetry, this);
    this.sendComment = __bind(this.sendComment, this);
    this.get = __bind(this.get, this);
    this.options = _utils.extend(this.options, this.constructor.defaultOptions);
    this.emit('connected');
    this._writeHeaders();
    if (this.options.compatibility) {
      this._compatibility();
    }
    if (this.options.retry) {
      this.sendRetry(this.options.retry);
    }
    if (this.options.keepAlive) {
      this._keepAlive();
    }
    if (!this.lastEventId) {
      this.lastEventId = this.req.headers['last-event-id'] || null;
    }
    if (this.lastEventId) {
      this.emit('reconnected');
    }
    this.res.once('close', function() {
      if (_this._keepAliveTimer) {
        clearTimeout(_this._keepAliveTimer);
      }
      return _this.emit('close');
    });
    this.emit('ready');
  }

  SSE.prototype.get = function(option) {
    var o;

    if (option in this.options) {
      return this.options[option];
    } else {
      throw new Error("Valid options are " + (((function() {
        var _results;

        _results = [];
        for (o in this.options) {
          _results.push(o);
        }
        return _results;
      }).call(this)).join(',')));
    }
  };

  SSE.prototype.set = function(option, value) {
    var o,
      _this = this;

    if (option in this.options) {
      this.options[option] = value;
      switch (option) {
        case 'retry':
          return this.sendRetry(this.options.retry);
        case 'keepAlive':
          return this.once('_keepAlive', function() {
            if (_this._keepAliveTimer) {
              clearTimeout(_this._keepAliveTimer);
            }
            return _this._keepAlive();
          });
        case 'compatibility':
          return this._compatibility();
      }
    } else {
      throw new Error("Valid options are " + (((function() {
        var _results;

        _results = [];
        for (o in this.options) {
          _results.push(o);
        }
        return _results;
      }).call(this)).join(',')));
    }
  };

  SSE.prototype.sendComment = function(comment) {
    return this.sendRaw(this.constructor.comment(comment));
  };

  SSE.prototype.sendRetry = function(time) {
    if (this.options.retry !== time) {
      this.options.retry = time;
    }
    return this.sendRaw(this.constructor.retry(time));
  };

  SSE.prototype.sendEvent = function(event) {
    return this.sendRaw(this.constructor.event(event));
  };

  SSE.prototype.sendId = function(id) {
    return this.sendRaw(this.constructor.id(id));
  };

  SSE.prototype.sendData = function(data) {
    return this.sendRaw(this.constructor.data(data));
  };

  SSE.prototype.sendRaw = function(data) {
    return this.res.write(data);
  };

  SSE.prototype._processAndSendMessage = function(message) {
    this.sendRaw(this.constructor.message(message));
    return this.sendRaw("\n");
  };

  SSE.prototype._dispatchMessage = function(message) {
    if (_utils.typeCheck('Object', message)) {
      return this._processAndSendMessage(message);
    } else if (_utils.typeCheck('String', message)) {
      return this.sendData(message);
    } else if (_utils.typeCheck('Array', message)) {
      return message.forEach(function(msg) {
        return this._dispatchMessage(msg);
      });
    } else {
      throw new Error("Unparsable message. (" + message + ")");
    }
  };

  SSE.prototype._writeHeaders = function() {
    var header, value, _ref, _results;

    _ref = this.constructor.headers();
    _results = [];
    for (header in _ref) {
      value = _ref[header];
      _results.push(this.res.setHeader(header, value));
    }
    return _results;
  };

  SSE.prototype._keepAlive = function() {
    var schedule,
      _this = this;

    schedule = function() {
      return setTimeout((function() {
        _this.sendComment("keepalive " + (Date.now()) + "\n\n");
        _this._keepAliveTimer = schedule();
        return _this.emit('_keepAlive');
      }), _this.options.keepAlive);
    };
    return this._keepAliveTimer = schedule();
  };

  SSE.prototype._compatibility = function() {
    this.res.setHeader('Transfer-Encoding', 'chunked');
    /* XDomainRequest (MSIE8, MSIE9)
    */

    this.sendComment(Array(2049).join(' '));
    /* Remy Sharp's Polyfill support.
    */

    if (this.req.headers['x-requested-with'] === 'XMLHttpRequest') {
      this.res.xhr = null;
    }
    if (url.parse(this.req.url, true).query.lastEventId) {
      return this.lastEventId = url.parse(this.req.url, true).query.lastEventId;
    }
  };

  SSE.prototype.toString = function() {
    var client;

    client = this.req.socket.address();
    return "<SSE " + client.address + ":" + client.port + " (" + client.family + ")>";
  };

  return SSE;

})(EventEmitter);

/* Aliases
*/


SSE.prototype.pub = SSE.prototype._dispatchMessage;

SSE.prototype.publish = SSE.prototype._dispatchMessage;

SSE.prototype.send = SSE.prototype._dispatchMessage;

module.exports = SSE;

/* Connect/Express middleware
*/


MIDDLEWARE_INSTANCE_PROPERTIES = ['sendComment', 'sendRetry', 'sendEvent', 'sendId', 'sendData', 'sendRaw', 'set', 'get', 'toString'];

middleware = function(req, res, options) {
  var callable, property, socket, _i, _len;

  callable = function(message) {
    return this.sse._socket.send(message);
  };
  socket = new SSE(req, res, options);
  for (_i = 0, _len = MIDDLEWARE_INSTANCE_PROPERTIES.length; _i < _len; _i++) {
    property = MIDDLEWARE_INSTANCE_PROPERTIES[_i];
    callable[property] = socket[property];
  }
  callable._socket = socket;
  return callable;
};

module.exports.middleware = function(options) {
  return function(req, res, next) {
    if (req.headers.accept === "text/event-stream") {
      if ((options != null ? options.compatibility : void 0) == null) {
        res.sse = middleware(req, res, options);
      }
    }
    if (next != null) {
      return next();
    }
  };
};
